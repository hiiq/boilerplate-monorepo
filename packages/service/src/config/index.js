import { Utils } from '@boilerplate-monorepo/common';
import 'dotenv/config';
import {
  always,
  both,
  complement,
  defaultTo,
  either,
  evolve,
  includes,
  isEmpty,
  mergeRight,
  pick,
  pipe,
  prop,
  propOr,
  startsWith,
  toLower,
  __,
} from 'ramda';
import { ProcessEnvKeys } from 'types/processEnv';

// eslint-disable-next-line no-process-env
const raw = pick(ProcessEnvKeys.values, process.env);

const defaults = {
  ACCESS_TOKEN_EXPIRY: '5m',
  ACCESS_TOKEN_SECRET: 'auth-token-secret',
  CAPTCHA_SECRET: '0x0000000000000000000000000000000000000000',
  CI: false,
  DATABASE_URL: '',
  DB_MIGRATION_SCHEMA: 'public',
  DB_MIGRATION_TABLE_NAME: 'migrations',
  DB_SCHEMA: 'main',
  DEBUG: `${raw.npm_package_name}*`,
  EMAIL_FROM_ADDRESS: `noreply@example.com`,
  EMAIL_FROM_NAME: 'No Reply',
  ENGINE_API_KEY: '',
  ENGINE_SCHEMA_TAG: 'local',
  HTTPS: true,
  NODE_ENV: 'development',
  PORT: 4000,
  REDIS_URL: '',
  REFRESH_TOKEN_EXPIRY: '7d',
  REFRESH_TOKEN_SECRET: 'refresh-token-secret',
  SHOW_CONFIG: false,
  SMTP_CONNECTION: '',
  UI_HOST_URI: 'https://local.host:3000',
};

const config = mergeRight(defaults, raw);

const normalize = pipe(prop('NODE_ENV'), defaultTo(''), toLower);

const isTest = always(
  pipe(normalize, includes(__, ['ci', 'jest', 'test']))(config) ||
    typeof test !== 'undefined'
);

const isDev = always(pipe(normalize, startsWith('dev'))(config));
const isProd = always(pipe(normalize, startsWith('prod'))(config));
const isSqlDebug = always(pipe(propOr('', 'DEBUG'), includes('sql')))(config);
const isUndefined = value => typeof value === 'undefined';
const isUndefinedOrEmpty = either(isUndefined, isEmpty);
const isNotDevAndEmpty = both(complement(isDev), isUndefinedOrEmpty);

const requiredEnvVars = [
  {
    failureTest: isUndefinedOrEmpty,
    name: 'DATABASE_URL',
  },
  {
    failureTest: isNotDevAndEmpty,
    name: 'REDIS_URL',
  },
];

const missingEnvVars = requiredEnvVars.filter(({ name, failureTest }) =>
  failureTest(config[name])
);

if (isProd() && missingEnvVars.length) {
  const messages = missingEnvVars.map(
    ({ name }) =>
      `${name} was not set. The application will not function correctly without it.`
  );

  // eslint-disable-next-line no-console
  console.error(messages.join('\n'));

  // eslint-disable-next-line no-process-exit
  process.exit(1);
}

const merged = pipe(
  evolve({
    HTTPS: Utils.toBool,
    PORT: Utils.toNumber,
    SHOW_CONFIG: Utils.toBool,
  }),
  mergeRight({
    isDev: isDev(),
    isProd: isProd(),
    isSqlDebug: isSqlDebug(),
    isTest: isTest(),
  })
)(config);

// eslint-disable-next-line no-console
merged.SHOW_CONFIG && console.log(JSON.stringify(merged, null, 2));

export { merged as config };
