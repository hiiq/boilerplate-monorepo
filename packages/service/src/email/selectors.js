const userToFormattedEmailAddress = ({ email, givenName, familyName }) =>
  `${givenName} ${familyName} <${email}>`;

export { userToFormattedEmailAddress };
