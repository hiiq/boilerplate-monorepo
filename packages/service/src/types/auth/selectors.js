import { AuthenticationError } from 'apollo-server-express';
import { isAfter } from 'date-fns';
import {
  JsonWebTokenError,
  sign,
  TokenExpiredError,
  verify,
} from 'jsonwebtoken';
import { log } from 'log';
import { logFactory } from 'log/logFactory';
import ms from 'ms';
import { InternalErrorMessage, PublicErrorMessage } from 'types/errorMessage';
import { ProcessEnvKeys } from 'types/processEnv';
import { Route } from 'types/route';
import { Enumeration } from './typedef';

const isPast = date => isAfter(new Date(), date);

const {
  [ProcessEnvKeys.ACCESS_TOKEN_EXPIRY]: ACCESS_TOKEN_EXPIRY,
  [ProcessEnvKeys.ACCESS_TOKEN_SECRET]: ACCESS_TOKEN_SECRET,
  NODE_ENV,
  [ProcessEnvKeys.REFRESH_TOKEN_EXPIRY]: REFRESH_TOKEN_EXPIRY,
  [ProcessEnvKeys.REFRESH_TOKEN_SECRET]: REFRESH_TOKEN_SECRET,
  // eslint-disable-next-line no-process-env
} = process.env;

const toRefreshToken = ({ email, id, role, tokenVersion, username }) => ({
  email,
  id,
  role,
  tokenVersion,
  username,
});

const toAccessToken = ({ email, id, role, username }) => ({
  email,
  id,
  role,
  username,
});

const encryptAccessToken = user =>
  sign(toAccessToken(user), ACCESS_TOKEN_SECRET, {
    expiresIn: ACCESS_TOKEN_EXPIRY,
  });

const encryptRefreshToken = user =>
  sign(toRefreshToken(user), REFRESH_TOKEN_SECRET, {
    expiresIn: REFRESH_TOKEN_EXPIRY,
  });

const decryptAccessToken = token => verify(token, ACCESS_TOKEN_SECRET);

const decryptRefreshToken = token => verify(token, REFRESH_TOKEN_SECRET);

// eslint-disable-next-line max-statements
const readAccessToken = req => {
  const debugLog = logFactory({
    method: 'readAccessToken',
    module: 'auth',
  });

  const bearerToken = req.headers[Enumeration.AUTHORIZATION_HEADER];

  if (!bearerToken) {
    return null;
  }

  try {
    const [, token] = bearerToken.split(' ');

    return decryptAccessToken(token);
  } catch (error) {
    if (error instanceof TokenExpiredError) {
      debugLog('📆 Access token expired');
    } else if (error instanceof JsonWebTokenError) {
      log.error(InternalErrorMessage.ACCESS_TOKEN_READ_ISSUE, error.message);
    } else {
      log.error(InternalErrorMessage.GENERIC, error);
    }

    return null;
  }
};

const readRefreshToken = req => {
  const debugLog = logFactory({
    method: 'readRefreshToken',
    module: 'auth',
  });

  const token = req.cookies[Enumeration.JWT_REFRESH_TOKEN_COOKIE_NAME];

  if (!token) {
    debugLog('🤷‍♂️ Refresh token not provided');

    return null;
  }

  try {
    return decryptRefreshToken(token);
  } catch (error) {
    log.error(InternalErrorMessage.FAILED_TO_DECRYPT_REFRESH_TOKEN, error);
  }

  return null;
};

const writeRefreshToken = (res, user) => {
  const debugLog = logFactory({
    method: 'writeRefreshToken',
    module: 'auth',
  });

  const maxAge = ms(REFRESH_TOKEN_EXPIRY);

  const options = {
    httpOnly: true,
    maxAge,
    path: Route.REFRESH_TOKEN,
    secure: NODE_ENV === 'production',
  };

  if (!user) {
    debugLog('🔥 Removing refresh token cookie value');
    res.cookie(Enumeration.JWT_REFRESH_TOKEN_COOKIE_NAME, '', options);

    return res;
  }

  const token = encryptRefreshToken(user);

  res.cookie(Enumeration.JWT_REFRESH_TOKEN_COOKIE_NAME, token, options);

  return res;
};

const authChecker = ({ context }, roles) => {
  const { user } = context;

  if (!user) {
    // No user present, restrict access
    throw new AuthenticationError(PublicErrorMessage.UNAUTHORIZED);
  }

  if (roles.length === 0) {
    // Valid user and no role restrictions, grant access
    return true;
  }

  if (roles.includes(user.role)) {
    // Grant access if the roles overlap
    return true;
  }

  // No roles matched, restrict access
  throw new AuthenticationError(PublicErrorMessage.UNAUTHORIZED);
};

const isUserActive = user => !user.deletedAt || !isPast(user.deletedAt);

export {
  authChecker,
  encryptAccessToken,
  isUserActive,
  readAccessToken,
  readRefreshToken,
  writeRefreshToken,
};
