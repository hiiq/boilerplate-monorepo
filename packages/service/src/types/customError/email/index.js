import { ApolloError } from 'apollo-server-express';
import { toSafeLog } from 'log/toSafeLog';
import { PublicErrorMessage } from 'types/errorMessage';
import { ErrorType } from 'types/errorType';

const toSafeError = toSafeLog;

const appendSafeError = props => ({
  ...props,
  toSafeError,
});

class EmailError extends ApolloError {
  constructor(properties) {
    super(
      PublicErrorMessage.FAILED_TO_SEND_OTP_EMAIL,
      ErrorType.FAILED_TO_SEND_OTP_EMAIL,
      appendSafeError(properties)
    );

    Object.defineProperty(this, 'name', { value: EmailError.name });
  }
}

export { EmailError };
