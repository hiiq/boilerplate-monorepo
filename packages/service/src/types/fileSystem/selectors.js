import { readFile as readFileFs } from 'fs';
import { promisify } from 'util';

const readFileAsync = promisify(readFileFs);

const readFile = (filePath) => readFileAsync(filePath, { flag: 'r' });

export { readFile };
