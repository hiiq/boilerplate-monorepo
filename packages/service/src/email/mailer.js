import { config } from 'config';
import { log } from 'log';
import nodemailer from 'nodemailer';
import url from 'url';

const { SMTP_CONNECTION } = config;

const make = () => {
  if (!SMTP_CONNECTION) {
    return {
      sendMail: props =>
        log.warn(
          'SMTP_CONNECTION was not set so emails will not be sent',
          props
        ),
    };
  }

  const smtpUri = url.parse(SMTP_CONNECTION);

  const { auth: authentication, hostname, port, protocol } = smtpUri;
  const isSecure = protocol.toLowerCase() !== 'smtp:';
  const [user, pass] = (authentication || '').split(':');
  const auth = pass && user ? { pass, user } : undefined;

  const options = {
    auth,
    host: hostname,
    port,
    secure: isSecure,
  };

  return nodemailer.createTransport(options);
};

const mailer = make();

export { mailer };
