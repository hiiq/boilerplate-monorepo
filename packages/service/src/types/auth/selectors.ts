import { AuthenticationError } from 'apollo-server-express';
import { isAfter } from 'date-fns';
import 'dotenv/config';
import { User } from 'entity/User';
import { Request, Response } from 'express';
import { JsonWebTokenError, sign, TokenExpiredError, verify } from 'jsonwebtoken';
import { log } from 'log';
import { AuthChecker } from 'type-graphql';
import { Context } from 'types/context';
import { InternalErrorMessage, PublicErrorMessage } from 'types/errorMessage';
import { ProcessEnvKeys } from 'types/processEnv';
import { RefreshToken } from 'types/refreshToken';
import { UserContext } from 'types/userContext';
import { UserRole } from 'types/userRole';
import { Enumeration } from './typedef';

const isPast = (date: Date) => isAfter(new Date(), date);

const toRefreshToken = ({
  email,
  id,
  role,
  tokenVersion,
  username,
}: User): RefreshToken => ({ email, id, role, tokenVersion, username });

const toAccessToken = ({ email, id, role, username }: User): UserContext => ({
  email,
  id,
  role,
  username,
});

const encryptAccessToken = (user: User) =>
  sign(toAccessToken(user), process.env[ProcessEnvKeys.ACCESS_TOKEN_SECRET]!, {
    expiresIn: process.env[ProcessEnvKeys.ACCESS_TOKEN_EXPIRY],
  });

const encryptRefreshToken = (user: User) =>
  sign(
    toRefreshToken(user),
    process.env[ProcessEnvKeys.REFRESH_TOKEN_SECRET]!,
    {
      expiresIn: process.env[ProcessEnvKeys.REFRESH_TOKEN_EXPIRY],
    }
  );

const decryptAccessToken = (token: string): UserContext =>
  verify(
    token,
    process.env[ProcessEnvKeys.ACCESS_TOKEN_SECRET]!
  ) as UserContext;

const decryptRefreshToken = (token: string): RefreshToken =>
  verify(
    token,
    process.env[ProcessEnvKeys.REFRESH_TOKEN_SECRET]!
  ) as RefreshToken;

const readAccessToken = (req: Request): UserContext | null => {
  const bearerToken = req.headers[Enumeration.AUTHORIZATION_HEADER];

  if (!bearerToken) {
    return null;
  }

  try {
    const [, token] = (bearerToken as string).split(' ');

    return decryptAccessToken(token);
  } catch (error) {
    if (error instanceof TokenExpiredError) {
      log.error(InternalErrorMessage.ACCESS_TOKEN_EXPIRED);
    } else if (error instanceof JsonWebTokenError) {
      log.error(InternalErrorMessage.ACCESS_TOKEN_READ_ISSUE, error.message);
    }
    else {
      log.error(InternalErrorMessage.GENERIC, error);
    }

    return null;
  }
};

const readRefreshToken = (req: Request) => {
  const token = req.cookies[Enumeration.JWT_REFRESH_TOKEN_COOKIE_NAME];

  if (!token) {
    log.error(InternalErrorMessage.REFRESH_TOKEN_COOKIE_NOT_FOUND);

    return null;
  }

  try {
    return decryptRefreshToken(token);
  } catch (error) {
    log.error(InternalErrorMessage.FAILED_TO_DECRYPT_REFRESH_TOKEN);
  }

  return null;
};

const writeRefreshToken = (res: Response, user: User) => {
  const token = encryptRefreshToken(user);

  res.cookie(Enumeration.JWT_REFRESH_TOKEN_COOKIE_NAME, token, {
    httpOnly: true,
  });
};

const authChecker: AuthChecker<Context, UserRole> = ({ context }, roles) => {
  const { user } = context;

  if (!user) {
    // No user present, restrict access
    throw new AuthenticationError(PublicErrorMessage.UNAUTHORIZED);
  }

  if (roles.length === 0) {
    // Valid user and no role restrictions, grant access
    return true;
  }

  if (roles.includes(user.role)) {
    // Grant access if the roles overlap
    return true;
  }

  // No roles matched, restrict access
  throw new AuthenticationError(PublicErrorMessage.UNAUTHORIZED);
};

const isUserActive = (user: User) => !user.deletedAt || !isPast(user.deletedAt);

export { authChecker, encryptAccessToken, isUserActive, readAccessToken, readRefreshToken, writeRefreshToken, };
