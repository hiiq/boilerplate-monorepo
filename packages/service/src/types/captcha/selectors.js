import { config } from 'config';
import https from 'https';
import querystring from 'querystring';

const { CAPTCHA_SECRET } = config;

const isTokenValid = token =>
  new Promise((resolve, reject) => {
    const data = querystring.stringify({
      response: token,
      secret: CAPTCHA_SECRET,
    });

    const options = {
      headers: {
        'content-length': Buffer.byteLength(data),
        'content-type': 'application/x-www-form-urlencoded',
      },
      host: 'hcaptcha.com',
      method: 'POST',
      path: '/siteverify',
    };

    const request = https.request(options, response => {
      response.setEncoding('utf8');

      let buffer = '';

      response
        .on('error', reject)
        .on('data', chunk => (buffer += chunk))
        .on('end', () => resolve(JSON.parse(buffer)));
    });

    request.on('error', reject);
    request.write(data);
    request.end();
  });

export { isTokenValid };
