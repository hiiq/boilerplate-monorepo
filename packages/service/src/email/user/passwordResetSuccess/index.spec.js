import { User } from '@boilerplate-monorepo/common';
import { config } from 'config';
import { bodyToHtml, userToFormattedEmailAddress } from 'email/selectors';
import { mailer as Mailer } from '../../mailer';
import { passwordResetSuccess } from './index';

const { EMAIL_FROM_ADDRESS, EMAIL_FROM_NAME } = config;

describe('passwordResetSuccess', () => {
  const user = User.apiExample();
  const input = { user };
  const from = `${EMAIL_FROM_NAME} <${EMAIL_FROM_ADDRESS}>`;
  const subject = 'ðŸ’š Your password has been reset';
  const to = userToFormattedEmailAddress(user);

  let sendMail = null;

  beforeEach(() => {
    sendMail = td.replace(Mailer, 'sendMail');
  });

  test('should send email', async () => {
    const body = await passwordResetSuccess(input);

    td.verify(
      sendMail({
        from,
        html: bodyToHtml(body),
        subject,
        to,
      })
    );
  });

  test('email body should match snapshot', async () => {
    const body = await passwordResetSuccess(input);

    expect(body).toMatchSnapshot();
  });
});
