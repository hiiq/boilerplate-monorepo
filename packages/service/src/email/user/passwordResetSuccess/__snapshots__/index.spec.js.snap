// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`passwordResetSuccess email body should match snapshot 1`] = `
"
    <div style=\\"
    border-spacing: 0;
    font-family: Arial,Helvetica,sans-serif;
    margin: 0 auto;
    max-width: 600px;
    text-align: center;
    width: 100%;
  \\">
      <h1>
        Hey there GIVEN_NAME, your password has been reset.
      </h1>
      <p>
        As a security measure we thought it would be a good idea to let you know that your password was reset a few seconds ago.
      </p>
      <p>
        <i>
          If you didnâ€™t ask to reset your password, please <a href=\\"https://local.host:3000/support\\">contact us</a> immediately.
        </i>
      </p>
    </div>
  "
`;
